File: backend/src/lib/entitlements.ts
import { pgQuery } from "./postgres.js";

/**
 * Free vs STAK+, decided in one place. A screen asks whether a feature is available,
 * never which plan the account is on, so packaging can change here without touching
 * any route or UI flow. Android mirrors this in data/Entitlements.kt.
 */
export type Plan = "free" | "plus";

/**
 * Features that go beyond the free core. Anything not listed is free. Limits that
 * aren't decided yet (the free "Adjust my interests" allowance) are added when they are.
 */
export type Feature =
	| "updates_history"
	| "taste_evolution";

const PLUS_ONLY: ReadonlySet<Feature> = new Set<Feature>(["updates_history", "taste_evolution"]);

export function planOf(raw: unknown): Plan {
	return raw === "plus" ? "plus" : "free";
}

export function hasFeature(plan: Plan, feature: Feature): boolean {
	return plan === "plus" || !PLUS_ONLY.has(feature);
}

/** The account's plan; an unknown account is free. */
export async function getPlan(uid: string): Promise<Plan> {
	const r = await pgQuery<{ plan: string | null }>(`select plan from users where uid = $1`, [uid]);
	return planOf(r.rows[0]?.plan);
}

File: android/app/src/main/java/com/stak/demo/data/Entitlements.kt
package com.stak.demo.data

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Free vs STAK+, decided in one place. A screen asks whether a feature is available,
 * never which plan the account is on, so packaging can change here without touching a
 * UI flow. Mirrors backend/src/lib/entitlements.ts; the server's plan (GET /api/me)
 * is the truth, and this copy only follows it.
 */
object Entitlements {
	enum class Plan { FREE, PLUS }

	/** Features beyond the free core. Anything not listed is free. */
	enum class Feature { UPDATES_HISTORY, TASTE_EVOLUTION }

	private val PLUS_ONLY = setOf(Feature.UPDATES_HISTORY, Feature.TASTE_EVOLUTION)
	private const val KEY = "entitlements.plan"

	var plan by mutableStateOf(Plan.FREE)
		private set

	fun has(feature: Feature): Boolean = plan == Plan.PLUS || feature !in PLUS_ONLY

	/** The plan the server reported; remembered so a cold start offline keeps it. */
	fun apply(raw: String?) {
		plan = if (raw == "plus") Plan.PLUS else Plan.FREE
		StakStore.putString(KEY, plan.name)
	}

	/** This account's last known plan, at account load. */
	fun load() {
		plan = if (StakStore.getString(KEY) == Plan.PLUS.name) Plan.PLUS else Plan.FREE
	}
}
